js/classes/entities/midground/weapon/sword.js

```diff
@@ -1,3 +1,201 @@
+        return 3;
+    };
+
+    /** The size of the Sword after the scale factor is applied. */
+    static get SCALED_SIZE() {
+        return Vector.multiply(Sword.SIZE, Sword.SCALE);
+    };
+
+    /** The maximum distance the Sword can be from Chad while jabbing before it begins retracting. */
+    static get MAX_OFFSET() {
+        return 30;
+    }
+
+    /** The offset from Chad's y-position applied to the Sword's y-position. */
+    static get Y_OFFSET() {
+        return 10;
+    }
+
+    /** The spritesheet containing all the Sword types. */
+    static get SPRITESHEET() {
+        return "./sprites/swords.png";
+    };
+
+    /** How fast the Sword moves while jabbing. */
+    static get JAB_SPEED() {
+        return 300;
+    };
+
+    /** How fast the Sword moves while retracting. */
+    static get RETRACT_SPEED() {
+        return 150;
+    };
+
+    /** Sword type 1 (rename later). */
+    static get TYPE_1() {
+        return 0;
+    }
+
+    /** Sword type 2 (rename later). */
+    static get TYPE_2() {
+        return 1;
+    }
+
+    /** Sword type 3 (rename later). */
+    static get TYPE_3() {
+        return 2;
+    }
+
+    /** Sword type 4 (rename later). */
+    static get TYPE_4() {
+        return 3;
+    }
+
+    /** Sword type 5 (rename later). */
+    static get TYPE_5() {
+        return 4;
+    }
+
+    /** The property table for Sword types. */
+    static get PROPERTY_TABLE() {
+        return {
+            [Sword.TYPE_1]: {
+                DAMAGE: 5
+            },
+            [Sword.TYPE_2]: {
+                DAMAGE: 10
+            },
+            [Sword.TYPE_3]: {
+                DAMAGE: 20
+            },
+            [Sword.TYPE_4]: {
+                DAMAGE: 40
+            },
+            [Sword.TYPE_5]: {
+                DAMAGE: 80
+            }
+        };
+    };
+
+    /**
+     * Helper method that checks the validity of the provided Sword type.
+     * 
+     * @param {number} type the type value to check
+     * @throws {Error} Will throw an error if type is invalid.
+     */
+    static checkType(type) { 
+        if (typeof type !== "number" || type % 1 !== 0 || type < 0 || type > 4) {
+            throw new Error("Invalid Sword type: please use a Sword member type (e.g. Sword.TYPE_1).");
+        }
+    }
+
+    /**
+     * Helper method. Gets the value of a property for the current Sword type.
+     * 
+     * @param {string} propertyName the name of the property
+     * @returns the value associated with propertyName for the Sword type
+     */
+    getProperty(propertyName) {
+        return Sword.PROPERTY_TABLE[this.type][propertyName];
+    };
+
+    /**
+     * Set the sword to another type.
+     * 
+     * @param {number} type the type of sword we're switching to
+     */
+    setType(type) {
+        this.type = type;
+        this.loadAnimations();
+    }
+
+    /** Update the Sword. */
+    update() {
+        if (GAME.user.jabbing && !this.isAttacking) {
```

Comment:
Can you explain what is happening in this if/else? In particular, I don't understand what's going on with the `offsetDirX`.

Reply:
The if-else updates the state of the Sword based on the progress of its animation as well as user input. Offset tracks the Sword's distance from its base position over the course of its animation, while `offsetDirX` is basically just a multiplier used to determine whether the offset is increasing or decreasing. Looking at it now I'm not sure that using a multiplier like this is the best approach and I can see why its purpose is unclear.